internal/wrphandlers/qos/qos.go

```diff
@@ -111,79 +109,62 @@ func (h *Handler) HandleWrp(msg wrp.Message) error {
 
 // serviceQOS is a long running goroutine that sends as many queued messages as possible,
 // where the highest QOS messages are prioritized.
-// serviceQOS starts when Handler.Start().
-// serviceQOS stops when Handler.Stop() is called.
-func (h *Handler) serviceQOS() {
+// Handler.Start starts serviceQOS.
+// Handler.Stop stops serviceQOS.
+func (h *Handler) serviceQOS(queue <-chan wrp.Message) {
 	var (
 		// Signaling channel from the handleWRP.
 		ready <-chan struct{}
 		// Channel for failed deliveries, re-enqueue message.
 		failedMsg <-chan wrp.Message
-		queue     <-chan wrp.Message
 	)
 
 	// create and manage the priority queue
 	pq := priorityQueue{maxQueueBytes: h.maxQueueBytes, maxMessageBytes: h.maxMessageBytes}
-	// handleWRP is promise pattern function
-	handleWRP := curryWRPHandler(h.next)
-
-	h.lock.Lock()
-	queue = h.queue
-	h.lock.Unlock()
-
 	for {
 		select {
 		case msg, ok := <-queue:
 			if !ok {
-				// Don't enqueue an empty wrp.Message{}
-				// Handler.Stop() has been called, both `queue` and `done` are closed.
+				// Handler.Stop has been called.
 				return
 			}
 
-			pq.Enqueue(msg)
-			if ready != nil {
-				// Previous handleWRP call has not finished, do nothing.
-			} else if top, ok := pq.Dequeue(); ok {
-				failedMsg, ready = handleWRP(top)
-			}
+			// ErrMaxMessageBytes errrors are ignored.
+			_ = pq.Enqueue(msg)
```

Comment:
if a ErrMaxMessageBytes error occurs, should qos still send a message (with a nil payload) if the message requires a transaction?

If so, what status code should we use?